packages/core/src/domain/session/sessionStoreOperations.ts

```diff
@@ -10,54 +9,89 @@ type Operations = {
+let currentLock: string | undefined
 
 export function processSessionStoreOperations(
   operations: Operations,
   sessionStoreStrategy: SessionStoreStrategy,
   numberOfRetries = 0
 ) {
-  const { retrieveSession, persistSession, clearSession } = sessionStoreStrategy
-  const lockEnabled = isLockEnabled()
-
+  const { retrieveSession, persistSession, clearSession, lockOptions } = sessionStoreStrategy
+  // If we do not have an ongoing operation, set the received operations
+  // as the current one
   if (!ongoingOperations) {
     ongoingOperations = operations
   }
+
+  // If we have a pending ongoing operation, we should buffer the received
+  // operation. It will be executed later.
   if (operations !== ongoingOperations) {
     bufferedOperations.push(operations)
     return
   }
-  if (lockEnabled && numberOfRetries >= LOCK_MAX_TRIES) {
+
+  // Lets look into the storage if we have a session. If we do not, the
+  // currentSession will be an empty object.
+  let currentSession = retrieveSession()
+
+  // If we were not able to fullfill our operation in the max amount of tries,
+  // discard it. It will not be executed.
+  if (lockOptions.enabled && numberOfRetries >= LOCK_MAX_TRIES) {
     next(sessionStoreStrategy)
+    // However - If we cannot get a lock within a reasonable amount of time, the lock
+    // may be in an inconsistent state (aborted process, manually modified, ...). We remove it.
+    delete currentSession.lock
+    persistSession(currentSession)
     return
   }
-  let currentLock: string
-  let currentSession = retrieveSession()
-  if (lockEnabled) {
-    // if someone has lock, retry later
-    if (currentSession.lock) {
+
+  if (lockOptions.enabled) {
+    // If a lock has been set, and it is not ours, we want to
+    // postpone the operation and retry later
+    // TODO: Check that this case is properly tested
+    if (currentSession.lock && currentSession.lock !== currentLock) {
       retryLater(operations, sessionStoreStrategy, numberOfRetries)
       return
     }
-    // acquire lock
-    currentLock = generateUUID()
-    currentSession.lock = currentLock
-    persistSession(currentSession)
-    // if lock is not acquired, retry later
-    currentSession = retrieveSession()
-    if (currentSession.lock !== currentLock) {
-      retryLater(operations, sessionStoreStrategy, numberOfRetries)
-      return
+
+    // Try to acquire our lock
+    if (currentSession.lock === undefined) {
+      currentLock = generateUUID()
+      currentSession.lock = currentLock
+      persistSession(currentSession)
+
+      // Here, typically for local storage, we must WAIT before retrieving the lock. Else, there is a chance another
+      // process can unknowingly write to the storage. This happens because of a latency to replicate the written value.
+      if (lockOptions.synchronizationLatency > 0) {
+        retryLater(operations, sessionStoreStrategy, numberOfRetries, lockOptions.synchronizationLatency)
```

Comment:
<!-- Suggestions are specific requests to improve the subject under review. It is assumed that we all want to do what's best, so these comments are never dismissed as “mere suggestions”, but are taken seriously.  -->
**💬 suggestion:** ‏`retry` is a bit confusing here, as you don't want to retry but actually continue the operation after a short while.

Maybe you could define a function that wraps everything after this `if`, and call it in a timeout:

```ts
      if (lockOptions.synchronizationLatency > 0) {
        setTimeout(onLockWritten, lockOptions.synchronizationLatency)
      } else {
        onLockWritten()
      }

  function onLockWritten() {
    if (lockOptions.enabled) {
      // If there is no latency, we retrieve our session to check we're still owner of our lock.
      // If the lock has changed, postpone the operation
      currentSession = retrieveSession()
      if (currentSession.lock !== currentLock) {
        retryLater(operations, sessionStoreStrategy, numberOfRetries)
        return
      }
    }

    let processedSession = operations.process(currentSession)

    if (lockOptions.enabled) {
      // If the lock has changed after process, retry later
      currentSession = retrieveSession()
      if (currentSession.lock !== currentLock) {
        retryLater(operations, sessionStoreStrategy, numberOfRetries)
        return
      }
    }

    ...
  }
```

Also, with this design, you wouldn't need to store the current lock value in the module scope (which adds complexity)